### PR TITLE
Allow filtering components in source structure

### PR DESCRIPTION
**Pull request by barreeeiroo**
This PR allows app developers to filter the source structure (_components tree_) by component type (all, visible and non-visible). This change is brought from Kodular sources.

image

### CODE REVIEW

**Comment by ewpatton** [comment]
@barreeeiroo @SusanRatiLane I have rebased this onto master now that the UI binder refactor has been merged. Tested both the classic and new forms. It works but I'm not super excited about having to pass the parent to the child so that the child can manipulate the parent. We'll have to think about that more.

**Review by SusanRatiLane — CHANGES_REQUESTED** [review]
We're hoping to have PRs approved on Monday. Let us know if you'll be unable to look at it.

File: appinventor/appengine/src/com/google/appinventor/client/explorer/SourceStructureExplorer.java [hunk 1]
```diff
@@ -174,8 +178,12 @@ public void onClick(ClickEvent event) {
...
     panel.add(scrollPanel);
-    panel.add(new Label());
-    panel.add(buttonPanel);
+    // TODO(diego@kodular.io): With App Inventor's current layout, as of now this is the only place to
+    //  render these buttons...
+    // if (includeButtonPanel) {
+      panel.add(new Label());
+      panel.add(buttonPanel);
+    // }
     panel.setCellHorizontalAlignment(buttonPanel, HorizontalPanel.ALIGN_CENTER);
```
[the remark this answers is not in the document]
Comment (SusanRatiLane) [reply appinventor/appengine/src/com/google/appinventor/client/explorer/SourceStructureExplorer.java:177]: > This was not commented out, the comment was there as in Kodular we have a toolbar with these buttons. There was the TODO comment to move them somewhere else.

Yes, it looks like I removed the buttons along with the comments. I'll put up a PR to fix the regression.

The intent of the TODO comment is still not clear to me.

[the remark this answers is not in the document]
Comment (barreeeiroo) [reply appinventor/appengine/src/com/google/appinventor/client/explorer/SourceStructureExplorer.java:177]: I'm working on it already, don't worry.

**Pull request merged**